### ISSUE TITLE
NestableNode.computeNextChild() is slow

### ISSUE THREAD

**Issue by markus-s24** [thymeleaf/thymeleaf#330]
When rendering pages with a lot of rows, NestableNode.computeNextChild() is really slow (rendering can take hours). The IdentityHashMap.containsKey() used in there consumes most of the computing time, because it is not just a lookup but needs an iteration too.

Isn't it possible to have an "already processed flag" at Node to keep its status to avoid costly iterations?

**Comment by jkuipers** [comment]
I can confirm this problem, it showed up during profiling of some large pages. 
It would help if a node could indicate if it inserted siblings before its own position so that if that's not the case, iterating over all the prior siblings again doesn't need to be done.

**Comment by danielfernandez** [comment]
Thanks. I've just updated 2.1.4-SNAPSHOT with an improvement that should importantly reduce the amount of times that `NestableNode.computeNextChild()` iterates on child nodes that have already been processed. This should mean a performance improvement in pages containing iterations on very large collections of elements.

Note however that `computeNextChild()` itself is the method that actually calls `process()` on each of the DOM tree nodes, so it is perfectly normal that it takes a lot execution time. What I have done with this modification is reducing its _own time_, i.e. the time that its code (and not that of `process()` or any other call made from it) takes to execute. As a lateral effect of this modification, `IdentityHashMap.containsKey()` should be called a much smaller number of times. In fact my tests after applying this patch show no significant difference between using the `IdentityHashMap` and not using it at all....

It would be great if you could do your performance checks again and tell me whether this has improved the situation for you too.

**Comment by danielfernandez** [comment]
Closing - this has been published as a part of 2.1.4

**Comment by danielfernandez** [comment]
Thymeleaf 2.1.4 has just been published, including this improvement.

**Comment by brdloush** [comment]
Hi, I'm wondering whether mentioned code can still be improved. We're still facing not-too-good performance with some of our templates (100-200ms just for rendering the html) and we traced the issue down to be related just to nested POJO objects. 

I've created a simple test project that's showing the poor performance - it can be cloned at following url: https://github.com/brdloush/thymeleaf-perf-issue.

As the profiler pointed out to computeNextChild(), I guess this might be related to this issue, so that's why I'm posting here.
